File: examples/vulnerable_samples/javascript/02_command_injection.js
const express = require('express');
const { exec } = require('child_process');

const app = express();

app.get('/ping', (req, res) => {
    const host = req.query.host || '127.0.0.1';
    exec(`ping -c 1 ${host}`, (err, stdout) => {
        if (err) return res.status(500).send('Ping failed');
        res.send(stdout);
    });
});

app.get('/backup', (req, res) => {
    const filename = req.query.file || 'default';
    exec(
        `tar -czf /backups/${filename}.tar.gz /data/${filename}`,
        (err, stdout) => {
            if (err) return res.status(500).send('Backup failed');
            res.send('Backup complete');
        }
    );
});

app.listen(3000);

File: examples/vulnerable_samples/python/02_command_injection.py
import os
import subprocess
from flask import Flask, request

app = Flask(__name__)


@app.route("/ping")
def ping_host():
    host = request.args.get("host", "127.0.0.1")
    # Attacker can submit ?host=1.1.1.1;rm -rf /
    os.system(f"ping -c 1 {host}")
    return f"Pinged {host}"


@app.route("/backup")
def backup_file():
    filename = request.args.get("file", "")
    # shell=True is the issue; filename is interpolated into a shell command
    result = subprocess.check_output(
        f"tar -czf /backups/{filename}.tar.gz /data/{filename}",
        shell=True,
    )
    return result


@app.route("/whoami")
def whoami():
    return subprocess.check_output(["whoami"]).decode().strip()


if __name__ == "__main__":
    app.run()
